### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by asteurer**

### CODE REVIEW

**Comment by asteurer** [comment]
I'm almost certain that the way I'm handling the package versions (latest and semvar) is wrong. Can I have one of you take  a look at the GH actions file and let me know if there's  a better way to do this?

**Comment by asteurer** [comment]
Also, this PR is solving https://github.com/spinkube/spin-operator/issues/277

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+name: Build and push redis package
+on:
+  push:
+    branches:
+      - main
+
+jobs:
+  build-on-push:
+    if: github.event_name == 'push'
```
Comment (bacongobbler) [inline .github/workflows/main.yaml:9]: this can be removed. Lines 2-5 already perform this logic.

File: .github/workflows/main.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    runs-on: ubuntu-latest
+    steps:     
+
+      - uses: actions/checkout@v3
```
Comment (bacongobbler) [inline .github/workflows/main.yaml:13]: The latest version available is v4: https://github.com/actions/checkout?tab=readme-ov-file#usage

```suggestion
      - uses: actions/checkout@v4
```

File: .github/workflows/main.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+name: Build and push redis package
+on:
+  push:
+    branches:
+      - main
+
+jobs:
+  build-on-push:
```
Comment (bacongobbler) [inline .github/workflows/main.yaml:8]:
```suggestion
  build-and-push:
```

File: .github/workflows/main.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    if: github.event_name == 'push'
+    runs-on: ubuntu-latest
+    steps:     
+
+      - uses: actions/checkout@v3
+
+      - name: install spin
+        uses: fermyon/actions/spin/setup@v1
+        with:
+          github_token: ${{ github.token }}
```
Comment (bacongobbler) [inline .github/workflows/main.yaml:18]: lines 17 and 18 can be removed.

File: .github/workflows/main.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+      - name: install go
+        uses: actions/setup-go@v3
```
Comment (bacongobbler) [inline .github/workflows/main.yaml:21]: The latest version is v5: https://github.com/actions/setup-go?tab=readme-ov-file#v5
```suggestion
        uses: actions/setup-go@v5
```

File: .github/workflows/main.yaml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: push package
+        run: |
+          RELEASE_VERSION=$(yq '.application.version' spin.toml) && \
+          spin registry push ghcr.io/${{ github.repository }}:$RELEASE_VERSION && \
```
Comment (bacongobbler) [inline .github/workflows/main.yaml:43]:
```suggestion
          spin registry push ghcr.io/${{ github.repository_owner }}/${{ github.repository }}:$RELEASE_VERSION && \
```

File: .github/workflows/main.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+          spin registry push ghcr.io/${{ github.repository }}:latest
```
Comment (bacongobbler) [inline .github/workflows/main.yaml:44]:
```suggestion
          spin registry push ghcr.io/${{ github.repository_owner }}/${{ github.repository }}:latest
```

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yaml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+      - name: login to GitHub container registry
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: push package
+        run: |
+          RELEASE_VERSION=$(yq '.application.version' spin.toml) && \
```
Comment (bacongobbler) [inline .github/workflows/main.yaml:42]: Have you looked at how spin-operator calculates the RELEASE_VERSION?

```
https://github.com/spinkube/spin-operator/blob/b8421d76181d24060bf2ed66ca78409de4e81b5b/.github/workflows/sample-apps.yaml#L24-L27
```

File: .github/workflows/main.yaml (hunk 7 above)
Comment (bacongobbler) [inline .github/workflows/main.yaml:44]:
```suggestion
          spin registry push ghcr.io/${{ github.repository_owner }}/${{ github.repository }}:latest
```

Comment (bacongobbler) [reply .github/workflows/main.yaml:44]: @endocrimes any opinions here on tagging `latest` for spinkube sample apps?

File: .github/workflows/main.yaml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: push package
+        run: |
+          RELEASE_VERSION=$(date +%Y%m%d-%H%M%S)-g$(git rev-parse --short HEAD) && \
+          spin registry push ghcr.io/${{ github.owner }}/${{ github.repository }}:$RELEASE_VERSION && \
+          spin registry push ghcr.io/${{ github.owner }}/${{ github.repository }}:latest
```
Comment (endocrimes) [inline .github/workflows/main.yaml:42]:
```suggestion
          spin registry push ghcr.io/${{ github.owner }}/${{ github.repository }}:$RELEASE_VERSION
```

I wouldn't bother publishing a `latest` image - it mostly results in documentation going stale quicker when you can't rely on pulling the same content.

File: main.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+		reqKey := r.Header.Get("x-key")
+		if reqKey == "" {
+			http.Error(w, "you must include the 'x-key' header in your request", http.StatusBadRequest)
+			return
+		}
+
+		if r.Method == "GET" {
+			value, err := rdb.Get(reqKey)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("no value found for key '%s'", reqKey), http.StatusBadRequest)
```
Comment (endocrimes) [inline main.go:39]:
```suggestion
				http.Error(w, fmt.Sprintf("no value found for key '%s'", reqKey), http.StatusNotFound)
```

File: main.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+			if err != nil {
+				http.Error(w, fmt.Sprintf("error deleting value for key '%w'", err), http.StatusInternalServerError)
+				return
+			}
+
+			w.WriteHeader(http.StatusOK)
+			return
+
+		} else {
+			http.Error(w, fmt.Sprintf("method %v is not supported, so please try again using 'GET' or 'PUT' for the HTTP method", r.Method), http.StatusBadRequest)
```
Comment (endocrimes) [inline main.go:73]:
```suggestion
			http.Error(w, fmt.Sprintf("method %q is not supported, so please try again using 'GET' or 'PUT' for the HTTP method", r.Method), http.StatusBadRequest)
```

`%q` will automatically quote the value, which is a little nicer

**Review by endocrimes — APPROVED** [review]
Thanks! - I'll give @bacongobbler a chance for a second review but LGTM

**Review by bacongobbler — APPROVED** [review]

**Pull request merged**